third_party/iap_jwt_verify_nginx/src/iap_jwt_verification.cc

```diff
@@ -0,0 +1,201 @@
+// modification, are permitted provided that the following conditions
+// are met:
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "src/iap_jwt_verification.h"
+
+#include <fstream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "openssl/bn.h"
+#include "openssl/bytestring.h"
+#include "openssl/ecdsa.h"
+#include "openssl/sha.h"
+#include "src/jwt.h"
+#include "src/key_util.h"
+
+namespace google {
+namespace cloud {
+namespace iap {
+
+// Let's refuse to even process JWTs larger than a certain size.
+constexpr size_t MAX_RAW_JWT_SIZE = 4096;
+
+// Allow for one minute of clock skew when verifying IAP JWT iat/exp values.
+constexpr uint64_t CLOCK_SKEW = 60;
+
+// Length of the elements of the finite field used for signing.
+constexpr unsigned int FINITE_FIELD_BYTE_SIZE = 32;
+
+bool iap_jwt_is_valid(const char *const raw_jwt,
+                      size_t raw_jwt_len,
+                      uint64_t now,
+                      char const *const expected_aud,
+                      size_t expected_aud_len,
+                      std::shared_ptr<iap_key_map_t> keys) {
+  if (raw_jwt == nullptr || raw_jwt_len > MAX_RAW_JWT_SIZE) {
+    return false;
+  }
+
+  std::unique_ptr<Jwt> jwt =
+      Jwt::parse_from_string(std::string(raw_jwt, raw_jwt_len));
+  if (jwt == nullptr) {
+    return false;
+  }
+
+  if (now < jwt->iat() - CLOCK_SKEW || now > jwt->exp() + CLOCK_SKEW) {
+    return false;
+  }
+
+  if (std::string(expected_aud, expected_aud_len) != jwt->aud()) {
+    return false;
+  }
+
+  if (keys == nullptr) {
+    return false;
+  }
+
+  if (!verify_iap_jwt_sig(*jwt, *keys)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool verify_iap_jwt_sig(const Jwt &jwt, const iap_key_map_t &keys) {
```

Comment:
Right, the algorithm is baked-in pretty hard. I'll add a comment calling this out, but I think it would be a waste of cycles to check the 'alg' value as things stand today.